### PR TITLE
#12 - Dashboard

### LINKED ISSUE #12: Dashboard
We should think of dashboard for student.

### PR DESCRIPTION
**Pull request by AmonDeShir**
Should close #12.

image
image
image

## Mobile
image
image

## Empty
image
image

### CODE REVIEW

File: app/Http/Controllers/QuizController.php [hunk 1]
```diff
@@ -85,4 +85,15 @@ public function createSubmission(Request $request, Quiz $quiz): RedirectResponse
...
+
+    public function assign(Request $request, Quiz $quiz)
+    {
+        $user = $request->user();
+        $quiz->assignedUsers()->attach($user);
+        $quiz->save();
+
+        return redirect()
+            ->back()
+            ->with("status", "Zostałeś przypisany do testu");
```
Comment (EwelinaSkrzypacz) [inline app/Http/Controllers/QuizController.php:97]:
```suggestion
            ->with("status", "Zostałeś przypisany do testu");
```

How about girls? "Zostałeś/aś"? Ormaybe something different?

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. When I click "Wyniki" I see 404 error - as I understand it's because we don't have rankings yet. My question is - is the situation secured such that if the user was not enrolled in the test, he should not enter the results page at all and the “Results” button should be disabled

image

2. Please create issue for dashboard for admin

File: app/Http/Controllers/QuizController.php (hunk 1 above)
Comment (AmonDeShir) [reply app/Http/Controllers/QuizController.php]: Done

**Comment by AmonDeShir** [comment]
> 1. When I click "Wyniki" I see 404 error - as I understand it's because we don't have rankings yet. My question is - is the situation secured such that if the user was not enrolled in the test, he should not enter the results page at all and the “Results” button should be disabled
> 
> image
> 
> 2. Please create issue for dashboard for admin

1. The history only shows completed tests in which the user participated. Even if a user tries to access the results of a test they did not take by using a direct URL, they should receive a 403 error. This functionality should be implemented in the rankings module.

2. We’ve decided that creating a dashboard for the admin doesn’t make sense with our current navigation. Instead of a dashboard, the admin will be redirected to the CRUD tests page.

**Review by EwelinaSkrzypacz — APPROVED** [review]

File: resources/js/Pages/User/Dashboard.vue [hunk 2]
```diff
@@ -1,3 +1,75 @@
...
+
+const started = computed(() => props.quizzes.filter(quiz => quiz.state == 'published' && !isClosed(quiz)))
+const scheduled = computed(() => props.quizzes.filter(quiz => quiz.state == 'locked'))
+const history = computed(() => props.submissions.filter(submission => submission.closed))
+</script>
+
 <template>
-  User Dashboard
+  <Head>
+    <title>Testy</title>
```
Comment (AleksandraKozubal) [inline resources/js/Pages/User/Dashboard.vue:25]: It's Testy or Konkursy?

File: resources/js/Pages/User/Dashboard.vue [hunk 3]
```diff
@@ -1,3 +1,75 @@
...
+
+    <Divider v-if="scheduled.length > 0 || started.length == 0">
+      <h1 class="font-bold text-xl text-primary text-center p-4 whitespace-nowrap">Nadchodzące konkursy</h1>
+    </Divider>
+    <div v-for="quiz in scheduled" :key="quiz.id" class="rounded-lg bg-white shadow border px-4 py-2 flex items-center justify-between mb-2">
+      <div>
+        <p class="font-semibold text-sm 2xs:text-base text-primary">{{ quiz.name }}</p>
+        <p class="text-xs py-1">{{ dayjs(quiz.scheduledAt).fromNow() }}</p>
+      </div>
+      <FormButton v-if="!quiz.isUserAssigned" class="min-w-24 text-center" small method="post" :href="`/quizzes/${quiz.id}/assign`" :options="{ preserveScroll: true }">
```
Comment (AleksandraKozubal) [inline resources/js/Pages/User/Dashboard.vue:50]: Maybe use named route?

File: resources/js/Pages/User/Dashboard.vue [hunk 4]
```diff
@@ -1,3 +1,75 @@
...
+        Zapisz się
+      </FormButton>
+      <FormButton v-else class="min-w-24 text-center" disabled small method="post" :href="`/quizzes/${quiz.id}/assign`">
```
Comment (AleksandraKozubal) [inline resources/js/Pages/User/Dashboard.vue:53]: Here also

**Review by AleksandraKozubal — COMMENTED** [review]
In the 'quiz_assignments' table columns `created_at` and `updated_at` is always empty

File: app/Models/Quiz.php [hunk 5]
```diff
@@ -102,6 +138,11 @@ public function isReadyToBePublished(): bool
         return $this->scheduled_at !== null && $this->duration !== null && $this->allQuestionsHaveCorrectAnswer();
     }
 
+    public function hasSubmissionsFrom(User $user): bool
+    {
+        return !$this->quizSubmissions->where("user_id", $user->id)->isEmpty();
```
[the remark this answers is not in the document]
Comment (AmonDeShir) [reply app/Models/Quiz.php:143]: done

File: database/migrations/2024_09_02_103155_create_quiz_subscriptions_table.php
[the remark this answers is not in the document]
Comment (AmonDeShir) [reply database/migrations/2024_09_02_103155_create_quiz_subscriptions_table.php:1]: done

File: app/Models/Quiz.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (AleksandraKozubal) [reply app/Models/Quiz.php:143]: in which commit?

File: resources/js/Pages/User/Dashboard.vue (hunk 4 above)
Comment (AmonDeShir) [reply resources/js/Pages/User/Dashboard.vue]: done

File: resources/js/Pages/User/Dashboard.vue (hunk 3 above)
Comment (AmonDeShir) [reply resources/js/Pages/User/Dashboard.vue]: done

File: resources/js/Pages/User/Dashboard.vue (hunk 2 above)
Comment (AmonDeShir) [reply resources/js/Pages/User/Dashboard.vue]: done

**Review by AleksandraKozubal — COMMENTED** [review]
Check if you give parameter everywhere it's needed, for example I cannot sign for future quiz bc of that:
image

**Review by AleksandraKozubal — COMMENTED** [review]
In db, table "quiz submissions' I am connected with quiz, but when i click "wyniki" i get 403. I should be able to see it.

**Review by AleksandraKozubal — COMMENTED** [review]
There is a wrong timezone set up, in db we have 11:18 instead of 13:18:>

File: resources/js/Pages/Auth/Verify-Email.vue [hunk 6]
```diff
@@ -1,15 +1,16 @@
 <script setup lang="ts">
```
Comment (EwelinaSkrzypacz) [inline resources/js/Pages/Auth/Verify-Email.vue:1]: Can we change filename to `VerifyEmail`?

**Review by EwelinaSkrzypacz — COMMENTED** [review]
1. I'm not sure if we really need ziggy. In the past, when project has this package, all routes were listed in the console. It's a bit dangerous (maybe it's changed, I don't know). But I will remove this package and make routes like we have on toby repo.
2. When I'm on dashboard, I have error (on other pages too)

image

File: resources/js/Pages/Auth/Verify-Email.vue (hunk 6 above)
Comment (AmonDeShir) [reply resources/js/Pages/Auth/Verify-Email.vue:1]: > Can we change filename to `VerifyEmail`?
done

Comment (AmonDeShir) [reply resources/js/Pages/Auth/Verify-Email.vue]: done

File: app/Http/Controllers/QuizSubmissionController.php [hunk 7]
```diff
@@ -17,4 +17,14 @@ public function show(QuizSubmission $quizSubmission): Response
 
         return Inertia::render("User/Quiz", ["submission" => QuizSubmissionResource::make($quizSubmission)]);
     }
+
+    public function result(QuizSubmission $quizSubmission): Response
+    {
+        $quizSubmission->load(["answerRecords.question.answers", "quiz"]);
+
+        return Inertia::render("User/QuizResult", [
```
Comment (AleksandraKozubal) [inline app/Http/Controllers/QuizSubmissionController.php:25]: It leads to 404 bc page is named differently

Comment (AmonDeShir) [reply app/Http/Controllers/QuizSubmissionController.php:25]: fixed

**Review by EwelinaSkrzypacz — COMMENTED** [review]
When I run `php artisan migrate:fresh --seed` command, sometimes, when I assign to test I see 403 error :eyes: 

image

**Review by AleksandraKozubal — COMMENTED** [review]
conflicts

**Comment by AmonDeShir** [comment]
> When I run `php artisan migrate:fresh --seed` command, sometimes, when I assign to test I see 403 error 👀

fixed

**Review by EwelinaSkrzypacz — DISMISSED** [review]

File: app/Http/Resources/AnswerRecordResource.php [hunk 8]
```diff
@@ -4,30 +4,53 @@
...
+    /**
+     * @return Collection<array>
+     */
+    protected function questionAnswersToArray(Question $question): Collection
+    {
+        if ($question->quiz->isRankingPublished) {
+            return $question->answers->map(fn(Answer $answer) => $this->getFullAnswer($answer))->collect();
+        }
+
+        return $question->answers->map(fn(Answer $answer) => $this->getMinimalAnswer($answer))->collect();
```
Comment (AleksandraKozubal) [inline app/Http/Resources/AnswerRecordResource.php:37]:
```suggestion
    return $question->answers->map(fn(Answer $answer) =>
        $question->quiz->isRankingPublished
            ? $this->getFullAnswer($answer)
            : $this->getMinimalAnswer($answer)
    )->collect();
```

File: app/Models/Quiz.php [hunk 9]
```diff
@@ -50,11 +55,42 @@ public function answers(): HasManyThrough
...
+    }
+
+    public function state(): Attribute
+    {
+        return Attribute::get(function (): string {
+            if ($this->isPublished) {
+                return "published";
+            }
+
+            return $this->isLocked ? "locked" : "unlocked";
```
Comment (AleksandraKozubal) [inline app/Models/Quiz.php:85]:
```suggestion
    return Attribute::get(fn(): string =>
        $this->isPublished ? "published" : ($this->isLocked ? "locked" : "unlocked")
    );
```

**Review by EwelinaSkrzypacz — APPROVED** [review]

**Review by AleksandraKozubal — APPROVED** [review]

**Pull request merged**